src/irmin-pack/pack_key.ml

```diff
@@ -0,0 +1,176 @@
+     computing hashes of structured values (e.g. inodes), so that this hashing
+     process is reproducible in different stores (w/ different offsets for the
+     values). *)
+  let equal a b =
+    let result = Hash.equal (to_hash a) (to_hash b) in
+    assert (
+      if not result then true
+      else
+        match (a.state, b.state) with
+        | Direct a, Direct b -> a.length = b.length
+        | _, _ -> true);
+    result
+  in
+  let compare a b = Hash.compare (to_hash a) (to_hash b) in
+  let pre_hash t f = Hash.pre_hash (to_hash t) f in
+  let encode_bin x f =
+    match (codec, x.state) with
+    | ( Hash_and_offset,
+        ( Direct { hash; offset; length = _ }
+        | Direct_unknown_length { hash; offset } ) ) ->
+        Hash.encode_bin hash f;
+        let buf = Bytes.create 8 in
+        Bytes.set_int64_be buf 0 (Int63.to_int64 offset);
+        f (Bytes.unsafe_to_string buf)
+    | Hash, Indexed t -> Hash.encode_bin t f
+    | Hash, (Direct { hash; _ } | Direct_unknown_length { hash; _ }) ->
+        Hash.encode_bin hash f
+    | Hash_and_offset, Indexed _ -> assert false
+  in
+  let decode_bin =
+    match codec with
+    | Hash_and_offset ->
+        fun buf pos_ref ->
+          let hash = Hash.decode_bin buf pos_ref in
+          let buf = Bytes.unsafe_of_string buf in
+          let offset = Bytes.get_int64_be buf !pos_ref |> Int63.of_int64 in
+          pos_ref := !pos_ref + 8;
+          { state = Direct_unknown_length { hash; offset } }
+    | Hash ->
+        fun buf pos_ref -> { state = Indexed (Hash.decode_bin buf pos_ref) }
+  in
+  let size_of =
+    match codec with
+    | Hash_and_offset -> Irmin.Type.Size.custom_static (Hash.encoded_size + 8)
+    | Hash -> Irmin.Type.Size.custom_static Hash.encoded_size
+  in
+  Irmin.Type.like (t hash_t) ~pre_hash ~equal ~compare
+    ~bin:(encode_bin, decode_bin, size_of)
+
+let v_direct ~hash ~offset ~length = { state = Direct { hash; offset; length } }
+
+let v_blindfolded ~hash ~offset =
+  { state = Direct_unknown_length { hash; offset } }
+
+module type S = sig
+  type hash
+
+  include S with type t = hash t and type hash := hash
+end
+
+module Make_no_repr (Hash : Irmin.Hash.S) = struct
+  type nonrec t = Hash.t t
+  type hash = Hash.t [@@deriving irmin ~of_bin_string]
+
+  let to_hash = to_hash
+
+  let to_offset t =
+    match t.state with
+    | Direct t -> t.offset
+    | Direct_unknown_length t -> t.offset
+    | Indexed _ -> assert false
+
+  let to_length t =
+    match t.state with
+    | Direct t -> t.length
+    | Direct_unknown_length _ | Indexed _ -> assert false
+
+  let null_offset = Int63.minus_one
+  let null_length = -1
+
+  let null =
+    let buf = String.make Hash.hash_size '\000' in
+    let hash =
+      match hash_of_bin_string buf with Ok x -> x | Error _ -> assert false
+    in
+    v_direct ~hash ~offset:null_offset ~length:null_length
+
+  let unfindable_of_hash hash =
```

Comment:
As discussed offline, we can reconsider this when refactoring the `layout` type in Inodes. Currently, this function is only used to enable an efficient implementation of `Inode.Portable.of_node`, and it may become redundant with a more expressive layout GADT. Until then, this nastiness is best kept to inodes (and not propagated to e.g. pattern matching on keys in `pack_store.ml`).

For now, I've added a comment to `val unfindable_of_hash` clarifying its intended use.